File: Implementations/C++/Algorithms/Sorting/countingSort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

typedef std::vector<int> vi;

/*
    * Counting sort details:
    The counting sort algorithm sorts an array by counting the number of occurrences of each element.
    It is a linear time algorithm in the worst case.

    * Best time complexity: O(n+k) where k is the range
    * Worst time complexity: O(n+k) where k is the range
    * Average time complexity: O(n+k) where k is the range

    * Space complexity: O(n+k) where k is the range

    * How it works: 
    * 1. Find minimum and maximum values in the array
    * 2. Create a count array of size (maximum - minimum + 1) with all values as 0
    * 3. For each element in the input array, increment the count array at the index of the element
    * 4. Change the count array to the cumulative sum of the elements
    * 5. Create a new array as a placeholder for the sorted array
    * 6. Find the index of each element of the original array in count array, and place the elements in the new array
*/
void countingSort(vi& arr, int n)
{
    int max = *max_element(arr.begin(), arr.end());
    int min = *min_element(arr.begin(), arr.end());
    int range = max - min + 1;
    vi countArr(range, 0);

    // storing count of each element
    for (int i = 0; i < n; i++) {
        countArr[arr[i] - min]++;
    }

    // changing countArr to store the cumulative sum of the elements
    // e.g. [1, 0, 2, 0 , 1], meaning the input array was [1, 3, 3, 5], based on the count of each element
    // to [1, 1, 3, 3, 4] based on the cumulative sum of each element
    for (int i = 1; i < range; i++) {
        countArr[i] += countArr[i - 1];
    }

    vi sortedArr(n);
    for (int i = n - 1; i >= 0; i--) {
        // Find the correct index in the sorted array, and place the element there
        sortedArr[countArr[arr[i] - min] - 1] = arr[i];
        // decrement the countArr count value
        countArr[arr[i] - min]--; 
    }
    arr = sortedArr;
}

int main()
{
    vi arr;
    std::cout << "Enter number of elements: ";
    int n; std::cin >> n;
    std::cout << "\n";

    std::cout << "Enter elements (with spaces as the delimiter):\n";
    for (int i=0; i < n; i++) {
        int x; std::cin >> x;
        arr.push_back(x);
    }

    std::cout << "\nSorted elements:\n";
    countingSort(arr, n);
    for (const auto& i: arr) std::cout << i << " ";
    return 0;
}

File: NYP-DSA-Assignment/src/test_algorithms/counting_sort.py
"""
Not implemented for the assignment since I already had radix sort which uses counting sort
but optimises its space complexity.
"""

def get_min_max(arr:list[int]) -> tuple[int, int]:
    """
    Returns the minimum and maximum values in the array
    
    Args:
        arr: list[int]
    """
    min_val = max_val = arr[0]

    for i in range(1, len(arr)):
        if (arr[i] < min_val):
            min_val = arr[i]
        if (arr[i] > max_val):
            max_val = arr[i]

    return min_val, max_val

def counting_sort(array:list[int], reverse:bool=False) -> None:
    """
    The counting sort algorithm sorts an array by counting 
    the number of occurrences of each element.
    It is a linear time algorithm in the worst case.
    
    Best time complexity: O(n+k) 
    Worst time complexity: O(n+k) 
    Average time complexity: O(n+k) 

    Space complexity: O(n+k) 
    where k is the range
    """
    n = len(array)

    minRange, maxRange = get_min_max(array)
    countRange = maxRange - minRange + 1

    outputArr = [0] * n

    # Initialize count array
    countArr = [0] * countRange

    # Store the count of each elements in count array
    for i in range(0, n):
        countArr[array[i] - minRange] += 1

    # Store the cummulative count
    if (not reverse):
        for i in range(1, countRange):
            countArr[i] += countArr[i - 1]
    else:
        for i in range(countRange - 2, -1, -1):
            countArr[i] += countArr[i + 1]

    # Find the index of each element of the original array in count array
    # place the elements in output array
    for i in range(n - 1, -1, -1):
        outputArr[countArr[array[i] - minRange] - 1] = array[i]
        countArr[array[i] - minRange] -= 1

    # Copy the sorted elements into original array
    for i in range(0, n):
        array[i] = outputArr[i]